mock/py/mockbuild/plugins/buildroot_lock.py

```diff
@@ -0,0 +1,92 @@
+        self.buildroot = buildroot
+        self.state = buildroot.state
+        self.conf = conf
+        self.inst_done = False
+        plugins.add_hook("postdeps", self.produce_lockfile)
+
+    def produce_lockfile(self):
+        """
+        Upon a request ('produce_lockfile' option set True), generate
+        the mock-build-environment.json file in resultdir.  The file describes
+        the Mock build environment, and the way to reproduce it.
+        """
+
+        filename = "buildroot.lock"
+        statename = "Generating the buildroot lockfile: " + filename
+        try:
+            with self.buildroot.uid_manager:
+                self.state.start(statename)
+                out_file = os.path.join(self.buildroot.resultdir, filename)
+                chrootpath = self.buildroot.make_chroot_path()
+
+                # Ḿimic the Koji Content Generator metadata fields:
+                # https://docs.pagure.org/koji/content_generator_metadata/#buildroots
+                #
+                # The query_packages() method below sorts its output according
+                # to _values_ of the queried RPM headers, so keep name-arch pair
+                # first to have the output sorted reasonably.
+                query_fields = ["name", "arch", "license", "version", "release",
+                                "epoch", "nvra", "sigmd5", "signature"]
+
+                def _executor(cmd):
+                    out, _ = self.buildroot.doOutChroot(cmd, returnOutput=True,
+                                                        returnStderr=False)
+                    return out
+
+                packages = query_packages(query_fields, chrootpath, _executor)
+                query_packages_location(packages, chrootpath, _executor)
+
+                data = {
+                    "version": "0",  # too-early stage for compatibility promises
+                    "buildroot": {
+                        "packages": packages,
+                    },
+                    # Try to keep this as minimal as possible.  If possible,
+                    # implement the config options as DEFAULTS in the
+                    # isolated-build.cfg, or in the
+                    # process_isolated_build_config() method.
+                    "config": {}
+                }
+                for cfg_option in [
+                    # These are hard-coded in the configuration file, but we
+                    # work with a single-config-for-all-arches now.
+                    "target_arch",
+                    "legal_host_arches",
+                    "dist",
+                    "package_manager",
+                    # At this point, we only support isolated builds iff
+                    # bootstrap_image_ready=True, so these two options are
+                    # useful for implementing "assertion" in the
+                    # process_isolated_build_config() method.
+                    "bootstrap_image",
```

Comment:
Bootstrap image would need also a sha256. Now it refers e.g. to "registry.fedoraproject.org/fedora:rawhide" which is always changing pointer.

Reply:
new field bootstrap.image_digest added

Reply:
and with the updated code, mock also asserts that the current digest equals the previous one